source/game/kart/KartCollide.cc

```diff
@@ -0,0 +1,278 @@
+            firstHitbox.worldPos(), firstHitbox.lastPos(), KCL_TYPE_IDK, &colInfo, &kclOut,
+            firstHitbox.radius(), 0);
+
+    CollisionData &collisionData = hitboxGroup->collisionData();
+
+    if (!collided) {
+        collisionData.speedFactor = 1.0f;
+        collisionData.rotFactor = 1.0f;
+        return;
+    }
+
+    collisionData.tangentOff = colInfo.tangentOff;
+
+    if ((kclOut & KCL_TYPE_FLOOR) != 0) {
+        collisionData.floor = true;
+        collisionData.floorNrm = colInfo.floorNrm;
+    }
+
+    collisionData.relPos = firstHitbox.worldPos() - pos();
+    collisionData.vel = _48;
+
+    processWheel(collisionData, firstHitbox, &colInfo, &kclOut);
+
+    if ((kclOut & KCL_TYPE_IDK) == 0) {
+        return;
+    }
+
+    Field::CollisionDirector::Instance()->findClosestCollisionEntry(&kclOut, KCL_TYPE_IDK);
+}
+
+void KartCollide::processWheel(CollisionData &collisionData, Hitbox &hitbox,
+        Field::CourseColMgr::CollisionInfo *colInfo, Field::KCLTypeMask *maskOut) {
+    processFloor(collisionData, hitbox, colInfo, maskOut, true);
+}
+
+void KartCollide::processFloor(CollisionData &collisionData, Hitbox & /*hitbox*/,
+        Field::CourseColMgr::CollisionInfo * /*colInfo*/, Field::KCLTypeMask *maskOut,
+        bool /*wheel*/) {
+    if ((*maskOut & KCL_TYPE_IDK) == 0) {
+        return;
+    }
+
+    if (!Field::CollisionDirector::Instance()->findClosestCollisionEntry(maskOut, KCL_TYPE_FLOOR)) {
+        return;
+    }
+
+    const Field::CollisionDirector::CollisionEntry *closestColEntry =
+            Field::CollisionDirector::Instance()->closestCollisionEntry();
+
+    if ((closestColEntry->attribute & 0x2000) == 0) {
+        m_notTrickable = true;
+    }
+
+    collisionData.intensity = (closestColEntry->attribute >> 0xb) & 3;
+    collisionData.closestFloorFlags = closestColEntry->typeMask;
+    collisionData.closestFloorSettings = (closestColEntry->attribute >> 5) & 7;
+
+    if ((*maskOut & KCL_TYPE_BIT(COL_TYPE_BOOST_RAMP)) == 0) {
+        m_notTrickable = true;
+    }
+}
+
+void KartCollide::applySomeFloorMoment(f32 down, f32 rate, CollisionGroup *hitboxGroup,
+        const EGG::Vector3f &forward, const EGG::Vector3f &nextDir, const EGG::Vector3f &speed,
+        bool b1, bool b2, bool b3) {
+    CollisionData &colData = hitboxGroup->collisionData();
+    if (!colData.floor) {
+        return;
+    }
+
+    f32 velDotFloorNrm = colData.vel.dot(colData.floorNrm);
+
+    if (velDotFloorNrm >= 0.0f) {
+        return;
+    }
+
+    EGG::Matrix34f rotMat;
+    rotMat.makeQ(dynamics()->mainRot());
+    EGG::Matrix34f tmp = rotMat.multiplyTo(dynamics()->invInertiaTensor());
+    tmp = tmp.multiplyTo(rotMat);
+
+    EGG::Vector3f crossVec = colData.relPos.cross(colData.floorNrm);
+    crossVec = tmp.multVector(crossVec);
+    crossVec = crossVec.cross(colData.relPos);
+
+    f32 idk = -velDotFloorNrm / (1.0f + colData.floorNrm.dot(crossVec));
```

Comment:
Gonna need a better name for this. Same with `dVar5` and `dVar7` later in this function.

Reply:
Renamed `idk` to something less stupid. The rest we can try to name as we become more familiar with the behavior over time.